lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+
+				commitConsensusMessage, err := commitMessage.ToConsensusMessage()
+				if err != nil {
+					return fmt.Errorf("transforming commit into consensus message: %w", err)
+				}
+
+				logger.Debugf("sending commit message: %v", commitMessage)
+
+				h.grandpaService.network.GossipMessage(commitConsensusMessage)
+				h.grandpaService.telemetry.SendMessage(telemetry.NewAfgFinalizedBlocksUpTo(
+					h.grandpaService.head.Hash(),
+					fmt.Sprint(h.grandpaService.head.Number),
+				))
+
+				logger.Debugf("round completed in %s", time.Since(start))
+				return nil
+
+			case alreadyFinalized:
+				logger.Debugf("round completed in %s", time.Since(start))
+				return nil
+			}
+		}
+	}
+}
+
+// actions that should take place accordingly to votes the
+// finalisation engine knows about
+type engineAction byte
+
+const (
+	determinePrevote engineAction = iota
+	determinePrecommit
+	alreadyFinalized
+	finalize
+)
+
+type finalizationEngine struct {
+	grandpaService *Service
+
+	timeoutStop time.Duration
+	stopCh      chan struct{}
+	engineDone  chan struct{}
+	actionCh    chan engineAction
+}
+
+func newFinalizationEngine(service *Service) *finalizationEngine {
+	return &finalizationEngine{
+		grandpaService: service,
+		timeoutStop:    5 * time.Second,
+		actionCh:       make(chan engineAction),
+		stopCh:         make(chan struct{}),
+		engineDone:     make(chan struct{}),
+	}
+}
+
+func (f *finalizationEngine) Stop() (err error) {
+	if f.stopCh == nil {
+		return nil
+	}
+
+	close(f.stopCh)
+	select {
+	case <-f.engineDone:
+	case <-time.After(f.timeoutStop):
+		return errTimeoutWhileStoping
+	}
+
+	f.stopCh = nil
+	close(f.actionCh)
+	return nil
+}
+
+func (f *finalizationEngine) Start() (err error) {
+	defer close(f.engineDone)
+
+	err = f.defineRoundVotes()
+	if err != nil {
+		return fmt.Errorf("defining round votes: %w", err)
+	}
+
+	err = f.finalizeRound()
+	if err != nil {
+		return fmt.Errorf("finalising round: %w", err)
+	}
+
+	return nil
+}
+
+func (f *finalizationEngine) defineRoundVotes() error {
+	gossipInterval := f.grandpaService.interval
+	determinePrevoteTimer := time.NewTimer(gossipInterval * 2)
+	determinePrecommitTimer := time.NewTimer(gossipInterval * 4)
+
+	var precommited bool = false
+
+	for !precommited {
+		select {
+		case <-f.stopCh:
+			if !determinePrevoteTimer.Stop() {
+				<-determinePrevoteTimer.C
+			}
+
+			if !determinePrecommitTimer.Stop() {
+				<-determinePrecommitTimer.C
+			}
+
+			return nil
+
+		case <-determinePrevoteTimer.C:
+			alreadyCompletable, err := f.grandpaService.checkRoundCompletable()
+			if err != nil {
+				return fmt.Errorf("checking round is completable: %w", err)
+			}
+
+			if alreadyCompletable {
+				f.actionCh <- alreadyFinalized
+				return nil
+			}
+
+			f.actionCh <- determinePrevote
+
+		case <-determinePrecommitTimer.C:
+			alreadyCompletable, err := f.grandpaService.checkRoundCompletable()
+			if err != nil {
+				return fmt.Errorf("checking round is completable: %w", err)
```

Comment:
the `determinePrevoteTimer` is always triggered before the `determinePrecommitTimer` so it is not needed to call the stop method here since the prevote timer was already called and will not be called anymore

Reply:
Indeed my bad!